### ISSUE TITLE
Neo4j 2.2.1: COST and RULE planners produce different results when using HEAD, TAIL, and LAST operations

### ISSUE THREAD

**Issue by bunkat** [neo4j/neo4j#4462]
I have a query that is calculating a minimum value across a set of related nodes. I was running into some odd results and it turns out that the two query planners produce different results for the following query:

```
  MATCH (p:project { id: "foobar" })
  WITH p

  MATCH (p)-[:MATERIAL]->(:materials)-[:MATERIAL]->(m:material)
  WITH m, [s IN (:schedule)<-[:SCHEDULE]-(:task)-[:MATERIAL]->(m) | HEAD(NODES(s))] AS schedules
  RETURN m, schedules
```

The COST planner decides that the `schedule` nodes are the HEAD elements while the RULE planner decides that the `task` nodes are the HEAD elements.  This causes queries to fail since unexpected nodes are being returned.

Rewriting the query this way makes both planners agree.

```
  WITH m, [s IN (m)<-[:MATERIAL]-(:task)-[:SCHEDULE]->(:schedule) | LAST(NODES(s))] AS schedules
```

Although the query is slightly ambiguous, the COST planner and RULE planner should always return the same results especially since the engine can decide to bail out of the COST planner based on previous statements in the query.

**Comment by bunkat** [comment]
Looks like the COST and RULE planners can't agree on even simpler examples:

```
[c IN (s)<-[:SCHEDULE]-(:week) | LAST(NODES(c))] AS contacts
```

COST likes LAST to return the `week` nodes, RULE likes HEAD. Re-writing the query slightly to this makes both planners agree that the HEAD nodes are the week nodes.

```
[c IN (:week)-[:SCHEDULE]->(s) | HEAD(NODES(c))] AS contacts
```

**Comment by jexp** [comment]
Hi Bill,
I tried it for a simple dataset and it worked for me.

can you try to run this and share the results:

```
match (s:schedule) 
return [c IN (s)<-[:SCHEDULE]-(:week) | NODES(c)] AS paths
```

@pontusmelke could that be related to the "getDegree" changes, where we would have to potentially reverse paths if we look from the other side first? left-to-right vs. right-to-left ?

**Comment by bunkat** [comment]
Yes, that query breaks as well between the two planners.

**RULE**

```
  PROFILE
  PLANNER RULE
  MATCH (s:schedule)
  RETURN [c IN (s)<-[:SCHEDULE]-(:week) | NODES(c)] AS paths LIMIT 1
```

![image](https://cloud.githubusercontent.com/assets/665885/7221457/8b59cdee-e6a0-11e4-8959-724a60d6fceb.png)

> [[:week, :schedule], [:week, :schedule], ... ] 

**COST**

```
  PROFILE
  PLANNER COST
  MATCH (s:schedule)
  RETURN [c IN (s)<-[:SCHEDULE]-(:week) | NODES(c)] AS paths LIMIT 1
```

![image](https://cloud.githubusercontent.com/assets/665885/7221470/052604da-e6a1-11e4-9238-cd919dc8e2d8.png)

> [[:schedule, :week], [:schedule, :week], ... ]

**Comment by spacecowboy** [comment]
Fixed in #4493
